### PR TITLE
New install algorithm

### PR DESCRIPTION
**Pull request by Morganamilo**
```I have replaced the old install and dependancy algorithms with a new
design that attemps to be more pacaur like. Mostly in minimizing user
input. Ask every thing first then do everything with no need for more
user input.

It is not yet fully complete but is finished enough so that it works,
should not fail in most cases and provides a base for more contributors
to help address the existing problems.

The new install chain is as follows:
	Source info about the provided targets
	Fetch a list of all dependancies needed to install targets
		I put alot of effort into fetching the dependancy tree
		while making the least amount of aur requests as
		possible. I'm actually very happy with how it turned out
		and yay wil now resolve dependancies noticably faster
		than pacaur when there are many aur dependancies.
	Install repo targets by passing to pacman
	Print dependancy tree and ask to confirm
	Ask to clean build if directory already exists
	Download all pkgbuilds
	Ask to edit all pkgbuilds
	Ask to continue with the install
	Download the sources for each packagebuild
	Build and install every package
		using -s to get repo deps and -i to install
	Ask to remove make dependancies

There are still a lot of things that need to be done for a fully working
system. Here are the problems I found with this system, either new or
existing:
	Formating
		I am not so good at formatting myself, I thought best to
		leave it until last so I could get feedback on how it
		should look and help implementing it.
	Dependancy tree
		The dependancy tree is usually correct although I have
		noticed times where it doesnt detect all the
		dependancies that it should. I have only noticed this
		when there are circular dependancies so i think this
		might be the cause. It's not a big deal currently
		because makepkg -i installed repo deps for us which
		handles the repo deps for us and will get the correct
		ones. So yay might not list all the dependancies. but
		they will get installed so I consider this a visual bug.
		I have yet to see any circular dependancies in the AUR
		so I can not say what will happend but I#m guessing that
		it will break.
	Versioned packages/dependencies
		Targets and dependancies with version constriants such
		as 'linux>=4.1' will not be checked on the aur side of
		things but will be checked on the repo side.
	Ignorepkg/Ignoregroup
		Currently I do not handle this in any way but it
		shouldn't be too hard to implement.
	Conflict checking
		This is not currently implemented either
	Split Paclages
		Split packages are not Handles properly. If we only
		specify one package so install from a split package
		makepkg -i ends up installing them all anyway. If we
		specify more than one (n) package it will actually build the
		package base n times and reinstall every split package
		n times.
	Makepkg
		To get things working I decided to keep using the
		makepkg -i method. I plan to eventually replace this
		with a pacman -U based method. This should allow passing
		args such as --dbpath and --config to aur packages
		aswell as help solve some problems such as the split
		packages.
	Clean build
		I plan to improve the clean build choice to be a little
		more smart and instead of check if the directory exists,
		check if the package is already build and if so skip the
		build all together.
```

Here is an example of the install process,  printing is non final and needs to be improved.
Note makepkg has been replaced by a dud program that prints the command issued as to simplify the output for this example.
 
image

image

This pr relates to #10 #27 #65 although might not directly fix them just yet. But they should be kept in mind. I will probably open an issue linking here and explaining the issue again after this is merged as a way to keep track of progress with this system.

### CODE REVIEW

**Comment by Jguer** [comment]
I like what I see 😀 , I'll review it as soon as possible.

**Comment by samosaara** [comment]
Wow a friend of mine @felipeac will have a stroke if this works, he always used to complain about how much he loved this pacaur feature and how [pacaur is now unmaintained](https://www.reddit.com/r/archlinux/comments/7k5suz/pacaur_now_unmaintained/), we can even end up filling a void

**Comment by Morganamilo** [comment]
@samosaara It's still a little broken right now, in fact I realised I kinda broke the number menu which needs to be fixed. I'll be away for the next day or two though sorry. Feel free to fix it yourself, after that the git branch should be good enough for every day use if you decided to use it.

**Comment by Morganamilo** [comment]
I took a look and it was simple enough to fix in five minutes so I went ahead and did it myself in #107

**Comment by maximbaz** [comment]
Wow this is astonishingly cool, great job @Morganamilo! 

> To get things working I decided to keep using the makepkg -i method. I plan to eventually replace this with a pacman -U based method...

So much looking forward to this!

File: install.go [hunk 1]
```diff
@@ -27,147 +38,216 @@ func install(parser *arguments) error {
...
+		fmt.Println()
+
+		fmt.Println()
+	
+		askCleanBuilds(dc.AurMake)
+		askCleanBuilds(dc.Aur)
+
+		fmt.Println()
+
+		if !continueTask("Proceed with download?", "nN") {
```
Comment (maximbaz) [inline install.go]: @Morganamilo this is a new confirmation comparing to the latest release, why did you add it? I was just asked about clean builds, if I wanted to cancel I would have simply pressed <kbd>Ctrl+C</kbd>. I'm happy that all questions are now being asked in the beginning, but I feel this is a question that can be avoided.

/cc @Jguer

**Comment by Morganamilo** [comment]
The printing was just done no get a feel for the flow and test the new system. it's all non final. Eventually I would like Y/N/A (yes/no/abort) on each question.

I plan to open an issue for the new install and another issue for how printing should go. I'm not home right now though. Feel free to open one for me. Would be nice to have a centralized place to discuss this.

**Comment by maximbaz** [comment]
Got it, thanks! Y/N/A would be the best, I agree. I'll let you create the issues yourself when you are back, no rush at all, I'm just trying this new code out to give you some early feedback 🙂
